api/client/cp.go

```diff
@@ -1,57 +1,253 @@
+// so we have to check for a `/` or `.` prefix. Also, in the case of a Windows
+// client, a `:` could be part of an absolute Windows path, in which case it
+// is immediately proceeded by a backslash.
+func splitCpArg(arg string) (container, path string) {
+	if filepath.IsAbs(arg) {
+		// Explicit local absolute path, e.g., `C:\foo` or `/foo`.
+		return "", arg
 	}
-	if statusCode == 404 {
-		return fmt.Errorf("No such container: %v", info[0])
+
+	parts := strings.SplitN(arg, ":", 2)
+
+	switch {
+	case len(parts) == 1:
+		// No `:` in the arg.
+		fallthrough
+	case strings.HasPrefix(parts[0], "."):
+		// Explicit local relative path.
+		return "", arg
+	default:
+		return parts[0], parts[1]
+	}
+}
+
+type containerPathStat struct {
+	Name  string      `json:"name"`
+	Path  string      `json:"path"`
+	Size  int64       `json:"size"`
+	Mode  os.FileMode `json:"mode"`
+	Mtime time.Time   `json:"mtime"`
+}
+
+func (cli *DockerCli) statContainerPath(containerName, path string) (containerPathStat, error) {
+	var stat containerPathStat
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(path)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/stat-path?%s", containerName, query.Encode())
+
+	response, _, statusCode, err := cli.call("GET", urlStr, nil, nil)
+	if err != nil {
+		return stat, err
+	}
+
+	defer response.Close()
+
+	if statusCode != 200 {
+		return stat, fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	err = json.NewDecoder(response).Decode(&stat)
+
+	return stat, err
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+
+	return archive.PreserveTrailingDotOrSeparator(absPath, localPath), nil
+}
+
+func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) (err error) {
+	if dstPath != "-" {
+		// Get an absolute destination path.
+		dstPath, err = resolveLocalPath(dstPath)
+		if err != nil {
+			return err
+		}
 	}
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(srcPath)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/archive-path?%s", srcContainer, query.Encode())
+
+	response, hdr, statusCode, err := cli.call("GET", urlStr, nil, nil)
 	if err != nil {
 		return err
 	}
 
-	hostPath := cmd.Arg(1)
-	if statusCode == 200 {
-		if hostPath == "-" {
-			_, err = io.Copy(cli.out, stream)
-		} else {
-			err = archive.Untar(stream, hostPath, &archive.TarOptions{NoLchown: true})
+	defer response.Close()
+
+	if statusCode != 200 {
+		return fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	if dstPath == "-" {
+		// Send the response to STDOUT.
+		_, err = io.Copy(os.Stdout, response)
+
+		return err
+	}
+
+	// Prepare source copy info.
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  hdr.Get("X-Docker-Container-Path-Is-Dir") == "true",
```

Comment:
We seem to be getting back to the issue where we are mixing up the goals of the `docker cp` command and the remote API. Maybe it makes sense to merge API first to avoid this confusion.

`docker cp` command does not have a directly corresponding endpoint in the API. So it doesn't make sense to even consider what is the actual tar moving on the wire when you run `docker cp`. For example the `docker cp` target path has different behavior depending on whether its a file or a directory, exists or not etc. The actual API endpoint `/containers/(id)/extract-to-dir?path=` has much simpler semantics - it requires path to be a directory, does not care if the path ends with a slash and simply extracts anything inside the tar into the target directory(no special rules about what is the first element in the archive etc.).

Even if in the future you would want to add support for something like `docker cp container:/foo - | tar -tv` the output of this command is only determined on the client side and does not depend on the semantics that would be decided for `/archive-path?path=`.

I think wildcards are much better solved with either `archive-path?path=/tmp&filter=foo*` or new completion endpoint `glob?path=/tmp/foo*` rather than expanding this to `archive-path?path=/tmp/foo*`. For a client developer the latter would again complicate things, meaning that I would have to check for `*` inside the path, possibly escape them, then check for the headers and then follow again some new set of rules when I discover that the result is actually a "multi result". The former ones would of course require some special handling from the client side for `docker cp` but as it is clearly visible from @duglin 's gist - these exeptions would be required anyway because the `cp -a` `source -> target` semantics don't work the same with wildcards.

It would be nice if I could get some critique for the idea of using `archive-path?path=/foo&dir=bool` instead of the `X-Docker-Container-Path-Is-Dir` header. Don't have to be in this thread if it gets too off-topic. Its much less important for me than my hope that API would only use cleaned paths but maybe still worth discussing a bit more.